contracts/FragToken.sol

```diff
@@ -49,24 +49,24 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable {
         _burn(account, amount);
     }
 
-    function lock(uint256 amount, bytes calldata signature, uint8 lock_period) external {
+    function lock(bytes calldata signature, uint256 amount, uint8 lock_period) external {
         require(amount > 0, "Amount must be greater than 0");
         require(lock_period >= 0 && lock_period <= 4, "Time lock period not allowed");
 
-        // make sure the signature is valid
-        bytes32 hash = ECDSA.toEthSignedMessageHash(
+        bytes32 digest = _hashTypedDataV4(
             keccak256(
-                abi.encodePacked(
-                    "FragLock",
+                abi.encode(
+                    keccak256(abi.encodePacked("Msg(string name,address sender,uint256 amount,uint8 lock_period)")),
+                    keccak256(abi.encodePacked("FragLock")),
                     msg.sender,
-                    uint64(block.chainid),
```

Comment:
Removed `chainid` from here because it is already used in the first part of the EIP-712 signature, not visible here, it's one in the EIP712 utility contract that this now uses.